src/backend/d3d12/DepthStencilStateD3D12.cpp

```diff
@@ -0,0 +1,101 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "backend/d3d12/DepthStencilStateD3D12.h"
+
+#include "common/BitSetIterator.h"
+
+namespace backend {
+namespace d3d12 {
+
+    static D3D12_STENCIL_OP StencilOp(nxt::StencilOperation op) {
+        switch (op) {
+            case nxt::StencilOperation::Keep:
+                return D3D12_STENCIL_OP_KEEP;
+            case nxt::StencilOperation::Zero:
+                return D3D12_STENCIL_OP_ZERO;
+            case nxt::StencilOperation::Replace:
+                return D3D12_STENCIL_OP_REPLACE;
+            case nxt::StencilOperation::IncrementClamp:
+                return D3D12_STENCIL_OP_INCR_SAT;
+            case nxt::StencilOperation::DecrementClamp:
+                return D3D12_STENCIL_OP_DECR_SAT;
+            case nxt::StencilOperation::Invert:
+                return D3D12_STENCIL_OP_INVERT;
+            case nxt::StencilOperation::IncrementWrap:
+                return D3D12_STENCIL_OP_INCR;
+            case nxt::StencilOperation::DecrementWrap:
+                return D3D12_STENCIL_OP_DECR;
+            default:
+                UNREACHABLE();
+        }
+    }
+
+    static D3D12_COMPARISON_FUNC ComparisonFunc(nxt::CompareFunction func) {
+        switch (func)
+        {
+            case nxt::CompareFunction::Always:
+                return D3D12_COMPARISON_FUNC_ALWAYS;
+            case nxt::CompareFunction::Equal:
+                return D3D12_COMPARISON_FUNC_EQUAL;
+            case nxt::CompareFunction::Greater:
+                return D3D12_COMPARISON_FUNC_GREATER;
+            case nxt::CompareFunction::GreaterEqual:
+                return D3D12_COMPARISON_FUNC_GREATER_EQUAL;
+            case nxt::CompareFunction::Less:
+                return D3D12_COMPARISON_FUNC_LESS;
+            case nxt::CompareFunction::LessEqual:
+                return D3D12_COMPARISON_FUNC_LESS_EQUAL;
+            case nxt::CompareFunction::Never:
+                return D3D12_COMPARISON_FUNC_NEVER;
+            case nxt::CompareFunction::NotEqual:
+                return D3D12_COMPARISON_FUNC_NOT_EQUAL;
+            default:
+                UNREACHABLE();
+        }
+    }
+
+    static D3D12_DEPTH_STENCILOP_DESC StencilOpDesc(backend::DepthStencilStateBase::StencilFaceInfo faceInfo) {
+        D3D12_DEPTH_STENCILOP_DESC desc;
+
+        desc.StencilFailOp = StencilOp(faceInfo.stencilFail);
+        desc.StencilDepthFailOp = StencilOp(faceInfo.depthFail);
+        desc.StencilPassOp = StencilOp(faceInfo.depthStencilPass);
+        desc.StencilFunc = ComparisonFunc(faceInfo.compareFunction);
+
+        return desc;
+    }
+
+    DepthStencilState::DepthStencilState(Device* device, DepthStencilStateBuilder* builder)
+        : DepthStencilStateBase(builder), device(device) {
+
+        // If you have anything other than Never, then enable depth testing
+        depthStencilDescriptor.DepthEnable = TRUE;
```

Comment:
It is difficult to answer if it would speed things up across all of the different hardware vendors and generations. There might be some theoretical boost but I think that binding / unbinding a depth buffer is probably already going to be a big hint that they follow.

After some internal discussion here I think that the NXT API needs to be explicit here. There are a few problems with implying the disable of depth checking in this way:

a) It is not super clear from the outside that you disable the hardware (including stencil and depth write, etc) just from setting one part of the depth stencil description.

b) If we try to do that here we will have two backends that are trying to extract real hardware state from a combination of state inputs.

Based on some of the previous remarks about this topic I think at a minimum we should revisit the API before we look for optimization opportunities here. Should we make an issue for this?